// ra-core/lib/controller/input/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useReferenceArrayInputController = exports.ReferenceInputController = exports.ReferenceArrayInputController = exports.getStatusForArrayInput = exports.getSelectedReferencesStatus = exports.getStatusForInput = void 0;
var ReferenceArrayInputController_1 = __importDefault(require("./ReferenceArrayInputController"));
exports.ReferenceArrayInputController = ReferenceArrayInputController_1.default;
var ReferenceInputController_1 = __importDefault(require("./ReferenceInputController"));
exports.ReferenceInputController = ReferenceInputController_1.default;
var useReferenceArrayInputController_1 = __importDefault(require("./useReferenceArrayInputController"));
exports.useReferenceArrayInputController = useReferenceArrayInputController_1.default;
var referenceDataStatus_1 = require("./referenceDataStatus");
Object.defineProperty(exports, "getStatusForInput", { enumerable: true, get: function () { return referenceDataStatus_1.getStatusForInput; } });
Object.defineProperty(exports, "getSelectedReferencesStatus", { enumerable: true, get: function () { return referenceDataStatus_1.getSelectedReferencesStatus; } });
Object.defineProperty(exports, "getStatusForArrayInput", { enumerable: true, get: function () { return referenceDataStatus_1.getStatusForArrayInput; } });
__exportStar(require("./useReferenceInputController"), exports);
