// ra-core/lib/form/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isRequired = exports.useWarnWhenUnsavedChanges = exports.ValidationError = exports.useSuggestions = exports.useInitializeFormWithRecord = exports.useInput = exports.useChoices = exports.sanitizeEmptyValues = exports.FormWithRedirect = exports.FormField = exports.FormDataConsumer = exports.addField = void 0;
var addField_1 = __importDefault(require("./addField"));
exports.addField = addField_1.default;
var FormDataConsumer_1 = __importDefault(require("./FormDataConsumer"));
exports.FormDataConsumer = FormDataConsumer_1.default;
var FormField_1 = __importDefault(require("./FormField"));
exports.FormField = FormField_1.default;
var FormWithRedirect_1 = __importDefault(require("./FormWithRedirect"));
exports.FormWithRedirect = FormWithRedirect_1.default;
var useInput_1 = __importDefault(require("./useInput"));
exports.useInput = useInput_1.default;
var ValidationError_1 = __importDefault(require("./ValidationError"));
exports.ValidationError = ValidationError_1.default;
var useInitializeFormWithRecord_1 = __importDefault(require("./useInitializeFormWithRecord"));
exports.useInitializeFormWithRecord = useInitializeFormWithRecord_1.default;
var sanitizeEmptyValues_1 = __importDefault(require("./sanitizeEmptyValues"));
exports.sanitizeEmptyValues = sanitizeEmptyValues_1.default;
var useChoices_1 = __importDefault(require("./useChoices"));
exports.useChoices = useChoices_1.default;
var useSuggestions_1 = __importDefault(require("./useSuggestions"));
exports.useSuggestions = useSuggestions_1.default;
var useWarnWhenUnsavedChanges_1 = __importDefault(require("./useWarnWhenUnsavedChanges"));
exports.useWarnWhenUnsavedChanges = useWarnWhenUnsavedChanges_1.default;
var FormField_2 = require("./FormField");
Object.defineProperty(exports, "isRequired", { enumerable: true, get: function () { return FormField_2.isRequired; } });
__exportStar(require("./validate"), exports);
__exportStar(require("./constants"), exports);
__exportStar(require("./FormContextProvider"), exports);
__exportStar(require("./FormContext"), exports);
__exportStar(require("./useFormContext"), exports);
__exportStar(require("./FormGroupContext"), exports);
__exportStar(require("./FormGroupContextProvider"), exports);
__exportStar(require("./useFormGroup"), exports);
__exportStar(require("./useFormGroupContext"), exports);
